daemon/cluster/executor/container/adapter.go

```diff
@@ -104,7 +104,10 @@ func (c *containerAdapter) pullImage(ctx context.Context) error {
 	go func() {
 		// TODO LCOW Support: This will need revisiting as
 		// the stack is built up to include LCOW support for swarm.
-		err := c.imageBackend.PullImage(ctx, c.container.image(), "", nil, metaHeaders, authConfig, pw)
+
+		// Make sure the image has a tag, otherwise it will pull all tags.
+		ref := reference.TagNameOnly(named)
+		err := c.imageBackend.PullImage(ctx, ref, nil, metaHeaders, authConfig, pw)
```

Comment:
The reason for the `docker-py` failure was that the `latest` tag wasn't enforced here, so this attempted to pull all tags from the repository (and that's not implemented yet).

Wondering why integration tests didn't pick this up...

Reply:
Would this be related to the trailing `:`?

Reply:
No, in this case the `:` was not involved at all. Passing an untagged reference to `PullImage` means "pull all tags". So if the image spec was just `busybox`, without `TagNameOnly` we would pull all tags instead of the `latest` one.

Reply:
oh! I think this would be because services always pin by digest, so `busybox` would be resolved to `busybox@digest`, and that's what's in the service's spec.

Reply:
But possibly `docker service create --no-resolve-image .....` could've run into that case if the image that's provide by the user doesn't have a `:tag` (but perhaps the CLI normalised that case)